### PR TITLE
Queryambiguous or fix

### PR DESCRIPTION
**Pull request by TabOg**

### CODE REVIEW

File: include/helib/partialMatch.h [hunk 1]
```diff
@@ -196,20 +196,14 @@ struct Expr;
 class ColNumber;
 
 /**
- * @brief An alias for a shared pointer to an `Expr` object.
+ * @brief A class for holding a shared pointer to an `Expr` object.
```
Comment (jlhcrawford) [inline include/helib/partialMatch.h]:
```suggestion
 * @brief A class wrapping a shared pointer to an `Expr` object.
```

File: include/helib/partialMatch.h [hunk 2]
```diff
@@ -196,20 +196,14 @@ struct Expr;
...
- * @return Shared pointer to the class `ColNumber`.
- **/
-inline std::shared_ptr<ColNumber> makeQueryExpr(int cl)
+class QueryExpr
 {
-  return std::make_shared<ColNumber>(cl);
-}
+  public:
+  std::shared_ptr<Expr> exp;
+  QueryExpr(std::shared_ptr<Expr> p) : exp(p) {}
```
Comment (jlhcrawford) [inline include/helib/partialMatch.h:209]: Doxygen comment for this constructor is missing.

File: tests/TestPartialMatch.cpp [hunk 3]
```diff
@@ -70,7 +70,8 @@ class TestPartialMatch : public ::testing::TestWithParam<BGVParameters>
                  addFrbMatrices(secretKey),
                  secretKey)),
       ea(context.getEA())
-  {}
+  {
+  }
```
Comment (jlhcrawford) [inline tests/TestPartialMatch.cpp]:
```suggestion
  {}
```

File: tests/TestPartialMatch.cpp [hunk 4]
```diff
@@ -712,34 +726,87 @@ TEST(TestPartialMatch, databaseLookupQueryAPIGeneratesMusAndTaus)
...
+      {qbExpand0, qbExpand1, qbExpand2, qbExpand3, qbExpand4, qbExpand5};
   long columns = 5;
 
-  helib::Query_t query = qb.build(columns);
-
-  std::vector<std::vector<long>> expected_Fs = {{0, 1, 2, 3, 4},
-                                                {0, 1, 2, 3, 4},
-                                                {0, 1, 2, 3, 4}};
-
+  std::vector<std::vector<std::vector<long>>> expected_Fs_vector(6);
```
Comment (jlhcrawford) [inline tests/TestPartialMatch.cpp]: Save the hardcoded value of 6 to a variable called `cases` or something. That way there's only one place to change it.

File: tests/TestPartialMatch.cpp [hunk 5]
```diff
@@ -712,34 +726,87 @@ TEST(TestPartialMatch, databaseLookupQueryAPIGeneratesMusAndTaus)
...
+  expected_taus_vector.push_back(expected_taus);
+
+  expected_taus = {{{1}, {0}, {0}, {0}, {0}},  // Only 0-th column
+                   {{0}, {1}, {1}, {0}, {0}},  // Either 1st or 2nd column
+                   {{0}, {1}, {0}, {1}, {0}}}; // Either 1st or 3rd column
+  expected_taus_vector.push_back(expected_taus);
+  std::vector<std::vector<long>> expected_mus_vector =
+      {{0, 0}, {0}, {0, 0}, {0, 0}, {0, 0, 0, 0}, {0, 0, 0}};
+
+  for (int j = 0; j < 6; j++) {
```
Comment (jlhcrawford) [inline tests/TestPartialMatch.cpp]: Change the 6 to the variable name you chose mentioned above.

Comment (jlhcrawford) [reply tests/TestPartialMatch.cpp]: Also should we check the sizes of expected_{Fs,taus,mus}_vector are the same?

**Review by jlhcrawford — CHANGES_REQUESTED** [review]

**Review by jlhcrawford — DISMISSED** [review]

File: include/helib/partialMatch.h [hunk 6]
```diff
@@ -240,30 +238,43 @@ class ColNumber : public Expr
...
-  ColNumber(int c) : column(c) {}
+  ColNumber(long c) : column(c) {}
 
 private:
-  int column;
+  long column;
 };
 
+/**
+ * @brief Utility function for creating a  QueryExpr for a specified column
```
Comment (jlhcrawford) [inline include/helib/partialMatch.h]:
```suggestion
 * @brief Utility function for creating a `QueryExpr` for a specified column
```

File: include/helib/partialMatch.h [hunk 7]
```diff
@@ -240,30 +238,43 @@ class ColNumber : public Expr
...
+ * in a query.
+ * @param cl The index of the column to be used in the query.
+ * @return QueryExpr with the specified `ColNumber` as the Expr.
```
Comment (jlhcrawford) [inline include/helib/partialMatch.h]:
```suggestion
 * @return `QueryExpr` with the specified `ColNumber` as the `Expr`.
```

File: include/helib/partialMatch.h [hunk 8]
```diff
@@ -311,29 +322,29 @@ class Or : public Expr
...
 /**
- * @brief Overloaded operator for creating a shared pointer to an AND
+ * @brief Overloaded operator for creating a `QueryExpr` to an `AND`
  * expression.
- * @param lhs Left operand of the AND expression.
- * @param rhs Right operand of the AND expression.
- * @return Shared pointer to the class `And`.
+ * @param lhs Left operand of the `AND` expression.
+ * @param rhs Right operand of the `AND` expression.
+ * @return QueryExpr with the `AND` as the Expr.
```
Comment (jlhcrawford) [inline include/helib/partialMatch.h]:
```suggestion
 * @return `QueryExpr` with the `AND` as the `Expr`.
```

File: include/helib/partialMatch.h [hunk 9]
```diff
@@ -311,29 +322,29 @@ class Or : public Expr
...
 /**
- * @brief Overloaded operator for creating a shared pointer to an OR
+ * @brief Overloaded operator for creating a `QueryExpr` of an `OR`
  * expression.
- * @param lhs Left operand of the OR expression.
- * @param rhs Right operand of the OR expression.
- * @return Shared pointer to the class `Or`.
+ * @param lhs Left operand of the `OR` expression.
+ * @param rhs Right operand of the `OR` expression.
+ * @return `QueryExpr` with the OR as the Expr.
```
Comment (jlhcrawford) [inline include/helib/partialMatch.h]:
```suggestion
 * @return `QueryExpr` with the OR as the `Expr`.
```

File: misc/psi/lookup/lookup.cpp [hunk 10]
```diff
@@ -100,8 +100,8 @@ int main(int argc, char *argv[])
   helib::QueryBuilder qb(a);
   helib::QueryBuilder qbAnd(a && b);
   helib::QueryBuilder qbOr(a || b);
-  helib::QueryBuilder qbExpand(a || (b && c));
-
+  helib::QueryBuilder qbExpand((a && b) || c);
```
Comment (jlhcrawford) [inline misc/psi/lookup/lookup.cpp]:
```suggestion
  helib::QueryBuilder qbExpand(a || (b && c));
```

**Review by jlhcrawford — CHANGES_REQUESTED** [review]

File: include/helib/partialMatch.h [hunk 11]
```diff
@@ -196,20 +196,18 @@ struct Expr;
...
-inline std::shared_ptr<ColNumber> makeQueryExpr(int cl)
+class QueryExpr
 {
-  return std::make_shared<ColNumber>(cl);
-}
+  public:
+  std::shared_ptr<Expr> exp;
+  /**
+   * @brief Constructor.
+   * @param p The shared pointer to wrap into a QueryExpr
```
Comment (jlhcrawford) [inline include/helib/partialMatch.h]:
```suggestion
   * @param p The `std::shared_ptr` to wrap into a `QueryExpr`.
```

**Review by jlhcrawford — CHANGES_REQUESTED** [review]

**Review by jlhcrawford — APPROVED** [review]

**Review by faberga — APPROVED** [review]

**Pull request merged**